src/transformers/generation/streamers.py

```diff
@@ -102,3 +103,93 @@ def end(self):
 
         # Print a newline (and the remaining text, if any)
         print(printable_text, flush=True)
+
+
+class IteratorStreamer(BaseStreamer):
+    """
+    Streamer that stores print-ready text in a queue, to be used by a downstream application as an iterator. This is
+    useful for applications that want to use the generated text in a non-blocking way (e.g. in an interactive Gradio
+    demo).
+
+    Parameters:
+        tokenizer (`AutoTokenizer`):
+            The tokenized used to decode the tokens.
+
+    Examples:
+
+        ```python
+        >>> from transformers import AutoModelForCausalLM, AutoTokenizer, IteratorStreamer
+        >>> from threading import Thread
+
+        >>> tok = AutoTokenizer.from_pretrained("gpt2")
+        >>> model = AutoModelForCausalLM.from_pretrained("gpt2")
+        >>> inputs = tok(["An increasing sequence: one,"], return_tensors="pt")
+        >>> streamer = IteratorStreamer(tok)
+
+        >>> # Run the generation in a separate thread, so that we can fetch the generated text in a non-blocking way.
+        >>> generation_kwargs = dict(inputs, streamer=streamer, max_new_tokens=20)
+        >>> thread = Thread(target=model.generate, kwargs=generation_kwargs)
+        >>> thread.start()
+        >>> generated_text = ""
+        >>> for new_text in streamer:
+        ...     generated_text += new_text
+        >>> generated_text
+        'An increasing sequence: one, two, three, four, five, six, seven, eight, nine, ten, eleven,'
+        ```
+    """
+
+    def __init__(self, tokenizer: "AutoTokenizer"):
+        self.tokenizer = tokenizer
+        self.token_cache = []
+        self.print_len = 0
+        self.queue = Queue()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """Gets the next item in the queue"""
+        value = self.queue.get()
+        if value is None:
+            raise StopIteration()
+        else:
+            return value
+
+    def put(self, value):
```

Comment:
Note: the logic in `put()` and `end()` is a near copy-paste from `TextStreamer` -- instead of printing, puts things in the queue. 

Most of the logic here is to work around model-specific tokenization quirks.